Investigate SQLAlchemy's nested transations

There is some evidence that suggests our nested transactions are not behaving properly. The ``moti_infill.py`` script is reporting that it's making insertions, however those insertions never actually appear in the database.

The transaction management in this package is "complicated". We haven't pinned our version of ``SQLAlchemy`` in the ``requirements.txt`` so there is a real possibility that the behaviour of nested transactions has changed out from under us. We need to:

1. Upgrade and pin to the latest stable version of ``SQLAlchemy`` and
1. Ensure that the nested transactions are behaving the way that they should

There appears to be a contradiction between what the documentation says and what we can write that works. (Caveat: Our code is not a minimal example, so there could be something non-obvious going on in it that confuses matters.)

[Documentation](http://docs.sqlalchemy.org/en/latest/orm/session_transaction.html#session-begin-nested) says:

> begin_nested() may be called any number of times, which will issue a new SAVEPOINT with a unique identifier for each call. For each begin_nested() call, a corresponding rollback() or commit() must be issued. (But note that if the return value is used as a context manager, i.e. in a with-statement, then this rollback/commit is issued by the context manager upon exiting the context, and so should not be added explicitly.)

1. Our original failing code:
```
            try:
                with sesh.begin_nested():
                    sesh.add(o)
                successes += 1
                log.debug("Inserted {}".format(o))
            except IntegrityError as e:
                log.debug("Skipped, already exists: {} {}".format(o, e))
                sesh.rollback()
                skips += 1
```
That code goes against the documentation by adding an explicit rollback (which is supposed to be provided by the context manager returned by `begin_nested`). And it fails because it provably (testably) rolls back all the objects inserted up to that point, but not those after.

2. Documentation-compliant failing code (remove supposedly spurious explicit rollback):
```
            try:
                with sesh.begin_nested():
                    sesh.add(o)
                successes += 1
                log.debug("Inserted {}".format(o))
            except IntegrityError as e:
                log.debug("Skipped, already exists: {} {}".format(o, e))
                skips += 1
```
That code fails with the following error message:
```
E                       sqlalchemy.exc.InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (psycopg2.IntegrityError) duplicate key value violates unique constraint "time_place_variable_unique"
E                       DETAIL:  Key (obs_time, history_id, vars_id)=(2012-01-02 01:00:00, 1, 1) already exists.
E                        [SQL: 'INSERT INTO crmp.obs_raw (obs_time, mod_time, datum, vars_id, history_id) VALUES (%(obs_time)s, %(mod_time)s, %(datum)s, %(vars_id)s, %(history_id)s) RETURNING crmp.obs_raw.obs_raw_id'] [parameters: {'obs_time': datetime.datetime(2012, 1, 2, 1, 0), 'history_id': 1, 'vars_id': 1, 'datum': -2.417, 'mod_time': datetime.datetime(2017, 11, 30, 0, 9, 58, 960640)}]
```

3. Failing alternative code that uses explicit commit and rollback, without a context manager:
```
            try:
                sesh.begin_nested()
                sesh.add(o)
                sesh.commit()
                successes += 1
                log.debug("Inserted {}".format(o))
            except IntegrityError as e:
                log.debug("Skipped, already exists: {} {}".format(o, e))
                sesh.rollback()
                skips += 1
```
That code also fails because it rolls back everything inserted before the rollback, but not the objects inserted after.

4. Working code that uses both context manager and explicit commit and rollback, contrary to recommendation in documenation:
```
            try:
                with sesh.begin_nested():
                    sesh.add(o)
                successes += 1
                sesh.commit()
                log.debug("Inserted {}".format(o))
            except IntegrityError as e:
                log.debug("Skipped, already exists: {} {}".format(o, e))
                sesh.rollback()
                skips += 1
```

Questions: WTF? What am I missing here?

--- a/.travis.yml
+++ b/.travis.yml
@@ -15,6 +15,4 @@ before_install:
 install:
   - pip install -i https://pypi.pacificclimate.org/simple/ -r requirements.txt -r test_requirements.txt
   - pip install .
-before_script:
-  - psql -U postgres -c "create extension postgis"
 script: py.test -v --tb=short tests

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,7 +1,6 @@
 dist: trusty
 language: python
 python:
-  - 3.3
   - 3.4
   - 3.5
   - 3.6

--- a/crmprtd/moti.py
+++ b/crmprtd/moti.py
@@ -145,7 +145,7 @@ def check_history(stn_id, sesh):
         with sesh.begin_nested():
             hist = History(station = stn)
             sesh.add(hist)
-    except Excetion as e:
+    except Exception as e:
         log.error('History_id could not be found or created for native_id {}'.format(stn_id), exc_info=True)
         raise e
     log.debug('Created history_id {}'.format(hist.id))

--- a/crmprtd/moti.py
+++ b/crmprtd/moti.py
@@ -98,6 +98,8 @@ def process_observation_series(sesh, os):
 
             o = Obs(time=t, datum=float(value), variable=var, history=hist)
 
+            # Regarding the following non-canonical but working code, see
+            # https://github.com/pacificclimate/crmprtd/issues/9#issuecomment-348042673
             try:
                 with sesh.begin_nested():
                     sesh.add(o)

--- a/crmprtd/moti.py
+++ b/crmprtd/moti.py
@@ -102,6 +102,7 @@ def process_observation_series(sesh, os):
                 with sesh.begin_nested():
                     sesh.add(o)
                 successes += 1
+                sesh.commit()
                 log.debug("Inserted {}".format(o))
             except IntegrityError as e:
                 log.debug("Skipped, already exists: {} {}".format(o, e))
@@ -109,6 +110,7 @@ def process_observation_series(sesh, os):
                 skips += 1
             except:
                 log.error("Failed to insert {}".format(o), exc_info=True)
+                sesh.rollback()
                 failures += 1
 
     return {'successes': successes, 'skips': skips, 'failures': failures}

--- a/test_requirements.txt
+++ b/test_requirements.txt
@@ -1,3 +1,3 @@
 pytest
-pytest-capturelog
+pytest-catchlog
 testing.postgresql

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -34,6 +34,7 @@ def postgis_session():
         sesh = sessionmaker(bind=engine)()
 
         logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)
+        sesh.execute('SET search_path TO crmp,public')
         yield sesh
 
 @pytest.yield_fixture(scope='function')
@@ -56,7 +57,7 @@ DECLARE
 BEGIN
     mystr = 'WITH stns_in_thresh AS (
     SELECT history_id, lat, lon, Geography(ST_Transform(the_geom,4326)) as p_existing, Geography(ST_SetSRID(ST_MakePoint('|| X ||','|| Y ||'),4326)) as p_new
-    FROM meta_history
+    FROM crmp.meta_history
     WHERE the_geom && ST_Buffer(Geography(ST_SetSRID(ST_MakePoint('|| X || ','|| Y ||'),4326)),'|| thres ||')
 )
 SELECT history_id, lat, lon, ST_Distance(p_existing,p_new) as dist
@@ -206,49 +207,115 @@ def moti_sawr7110_xml():
 <cmml xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="..\Schema\CMML.xsd" version="2.01">
   <head>
     <product operational-mode="official">
-      <title>Observation from BC Meteorological Stations
-      </title>
-      <field>meteorological
-      </field>
-      <category>observation
-      </category>
-      <creation-date refresh-frequency="PT1H">2013-11-29T13:22:57-08:00
-      </creation-date>
+      <title>Observation from BC Meteorological Stations</title>
+      <field>meteorological </field>
+      <category>observation</category>
+      <creation-date refresh-frequency="PT1H">2013-11-29T13:22:57-08:00</creation-date>
     </product>
     <source>
       <production-center>British Columbia Ministry of Transportation
-	<sub-center>AWP
-	</sub-center>
+        <sub-center>AWP</sub-center>
       </production-center>
     </source>
   </head>
   <data>
     <observation-series>
       <origin type="station">
-	<id type="client">11091
-	</id>
-	<id type="network">BC_MoT_11091
-	</id>
+        <id type="client">11091 </id>
+        <id type="network">BC_MoT_11091 </id>
       </origin>
       <observation valid-time="2012-01-01T00:00:00-08:00">
-	<pressure index="1" type="atmospheric">
-	  <value units="mb">964
-	  </value>
-	</pressure>
-	<temperature index="1" type="air-temperature">
-	  <value units="degC">-2.368
-	  </value>
-	</temperature>
+        <pressure index="1" type="atmospheric">
+          <value units="mb">964</value>
+        </pressure>
+        <temperature index="1" type="air-temperature">
+          <value units="degC">-2.368</value>
+        </temperature>
       </observation>
       <observation valid-time="2012-01-01T01:00:00-08:00">
-	<temperature index="1" type="air-temperature">
-	  <value units="degC">-2.417
-	  </value>
-	</temperature>
-	<temperature index="1" type="dew-point">
-	  <value units="degC">-4
-	  </value>
-	</temperature>
+        <temperature index="1" type="air-temperature">
+          <value units="degC">-2.417</value>
+        </temperature>
+        <temperature index="1" type="dew-point"> 
+          <value units="degC">-4</value>
+        </temperature>
+      </observation>
+    </observation-series>
+  </data>
+</cmml>''')
+
+@pytest.fixture(scope='module')
+def moti_sawr7110_xml_2a():
+    """No duplicates"""
+    return fromstring(b'''<?xml version="1.0" encoding="ISO-8859-1" ?>
+<cmml xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="..\Schema\CMML.xsd" version="2.01">
+  <head>
+    <product operational-mode="official">
+      <title>Observation from BC Meteorological Stations</title>
+      <field>meteorological </field>
+      <category>observation</category>
+      <creation-date refresh-frequency="PT1H">2013-11-29T13:22:57-08:00</creation-date>
+    </product>
+    <source>
+      <production-center>British Columbia Ministry of Transportation
+        <sub-center>AWP</sub-center>
+      </production-center>
+    </source>
+  </head>
+  <data>
+    <observation-series>
+      <origin type="station">
+        <id type="client">11091 </id>
+        <id type="network">BC_MoT_11091 </id>
+      </origin>
+      <observation valid-time="2012-01-02T01:00:00-08:00">
+        <temperature index="1" type="air-temperature">
+          <value units="degC">-2.417</value>
+        </temperature>
+      </observation>
+    </observation-series>
+  </data>
+</cmml>''')
+
+@pytest.fixture(scope='module')
+def moti_sawr7110_xml_2b():
+    """Duplicates observations in 2a, plus non-duplicate observations
+    before and after the duplicates."""
+    return fromstring(b'''<?xml version="1.0" encoding="ISO-8859-1" ?>
+<cmml xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="..\Schema\CMML.xsd" version="2.01">
+  <head>
+    <product operational-mode="official">
+      <title>Observation from BC Meteorological Stations</title>
+      <field>meteorological </field>
+      <category>observation</category>
+      <creation-date refresh-frequency="PT1H">2013-11-29T13:22:57-08:00</creation-date>
+    </product>
+    <source>
+      <production-center>British Columbia Ministry of Transportation
+        <sub-center>AWP</sub-center>
+      </production-center>
+    </source>
+  </head>
+  <data>
+    <observation-series>
+      <origin type="station">
+        <id type="client">11091 </id>
+        <id type="network">BC_MoT_11091 </id>
+      </origin>
+      <observation valid-time="2012-01-01T01:00:00-08:00">
+        <temperature index="1" type="air-temperature">
+          <value units="degC">-2.417</value>
+        </temperature>
+      </observation>
+      <observation valid-time="2012-01-02T01:00:00-08:00">
+        <temperature index="1" type="air-temperature">
+          <value units="degC">-2.417</value>
+        </temperature>
+      </observation>
+      <observation valid-time="2012-01-03T01:00:00-08:00">
+        <temperature index="1" type="air-temperature">
+          <value units="degC">-2.417</value>
+        </temperature>
       </observation>
     </observation-series>
   </data>
@@ -261,7 +328,7 @@ def moti_sawr7110_new_station():
   <data>
     <observation-series>
       <origin type="station">
-	    <id type="client">11092</id>
+        <id type="client">11092</id>
       </origin>
     </observation-series>
   </data>

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -82,7 +82,7 @@ def test_session(crmp_session, caplog):
     '''
     Yields a PostGIS enabled session with CRMP schema and test data
     '''
-    caplog.setLevel(logging.ERROR, logger='sqlalchemy.engine')
+    caplog.set_level(logging.ERROR, logger='sqlalchemy.engine')
 
     moti = Network(name='MoTIe')
     ec = Network(name='EC_raw')
@@ -151,7 +151,7 @@ def ec_session(crmp_session, caplog):
     '''
     Yields a PostGIS enabled session with CRMP schema and test data
     '''
-    caplog.setLevel(logging.ERROR, logger='sqlalchemy.engine')
+    caplog.set_level(logging.ERROR, logger='sqlalchemy.engine')
 
     ec = Network(name='EC_raw')
     crmp_session.add(ec)

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -270,7 +270,7 @@ def moti_sawr7110_xml_2a():
       </origin>
       <observation valid-time="2012-01-02T01:00:00-08:00">
         <temperature index="1" type="air-temperature">
-          <value units="degC">-2.417</value>
+          <value units="degC">2</value>
         </temperature>
       </observation>
     </observation-series>
@@ -304,17 +304,17 @@ def moti_sawr7110_xml_2b():
       </origin>
       <observation valid-time="2012-01-01T01:00:00-08:00">
         <temperature index="1" type="air-temperature">
-          <value units="degC">-2.417</value>
+          <value units="degC">1</value>
         </temperature>
       </observation>
       <observation valid-time="2012-01-02T01:00:00-08:00">
         <temperature index="1" type="air-temperature">
-          <value units="degC">-2.417</value>
+          <value units="degC">2</value>
         </temperature>
       </observation>
       <observation valid-time="2012-01-03T01:00:00-08:00">
         <temperature index="1" type="air-temperature">
-          <value units="degC">-2.417</value>
+          <value units="degC">3</value>
         </temperature>
       </observation>
     </observation-series>

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -7,7 +7,7 @@ import logging, logging.config
 import yaml
 import pytest
 import sqlalchemy
-from sqlalchemy.schema import DDL
+from sqlalchemy.schema import DDL, CreateSchema
 from sqlalchemy.orm import sessionmaker
 from lxml.etree import parse, fromstring, XSLT
 import testing.postgresql
@@ -30,6 +30,7 @@ def postgis_session():
     with testing.postgresql.Postgresql() as pg:
         engine = sqlalchemy.create_engine(pg.url())
         engine.execute("create extension postgis")
+        engine.execute(CreateSchema('crmp'))
         sesh = sessionmaker(bind=engine)()
 
         logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)

--- a/tests/test_ec.py
+++ b/tests/test_ec.py
@@ -442,7 +442,7 @@ def test_insert_duplicate_obs(ec_session, et, hid, vname, vid):
 
 def test_process_xml(ec_session, caplog):
     import logging
-    caplog.setLevel(logging.INFO)
+    caplog.set_level(logging.INFO)
 
     from ec_data import hourly_bc_2016061115, hourly_bc_2016061116
 

--- a/tests/test_moti.py
+++ b/tests/test_moti.py
@@ -5,7 +5,7 @@ import pytz
 from lxml.etree import fromstring, parse, XSLT
 import pytest
 
-from pycds import Obs
+from pycds import Obs, Station, History
 from crmprtd.moti import process, url_generator, slice_timesteps
 
 bctz = pytz.timezone('America/Vancouver')
@@ -28,6 +28,41 @@ def test_catch_duplicates(test_session, moti_sawr7110_xml):
     test_session.commit()
     rv = process(test_session, moti_sawr7110_xml)
     assert rv == {'failures': 0, 'successes': 0, 'skips': 4}
+
+
+def test_duplicates(test_session, moti_sawr7110_xml_2a, moti_sawr7110_xml_2b):
+    """Test that duplicates are skipped, but that skipping them does not
+    cause preceding or following non-duplicate inserts to be abandoned."""
+
+    # Insert initial observation set, no duplicates
+    try:
+        rv = process(test_session, moti_sawr7110_xml_2a)
+    except Exception as e:
+        assert False, \
+            "Didn't expect to get an exception on first insert\n{}".format(e)
+    # Check the claim about what happened
+    assert rv == {'failures': 0, 'successes': 1, 'skips': 0}
+
+    # Insert next observation set, with duplicate of first
+    try:
+        rv = process(test_session, moti_sawr7110_xml_2b)
+    except Exception as e:
+        assert False, \
+            "Didn't expect to get an exception on second insert\n{}".format(e)
+    # Check the claim about what happened
+    assert rv == {'failures': 0, 'successes': 2, 'skips': 1}
+
+    # Check that all non-duplicates actually got commited
+    # to the database and not rolled back due to the duplicate
+    test_session.commit()
+    q = (
+        test_session.query(Obs)
+        .join(History)
+        .join(Station)
+        .filter(Station.native_id == '11091')
+    )
+    assert q.count() == 3
+
 
 xml = {'no_obs': b'''<?xml version="1.0" encoding="ISO-8859-1" ?>
 <cmml xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="..\Schema\CMML.xsd" version="2.01">

--- a/tests/test_moti.py
+++ b/tests/test_moti.py
@@ -281,6 +281,6 @@ def test_unknown_var(test_session, caplog):
     assert n_obs_before == n_obs_after
 
     t = 'Could not find variable temperature, tree-temperature, celsius in the database. Skipping this observation.'
-    assert t in caplog.text()
+    assert t in caplog.text
 
     # TODO: need to actually check log warning

--- a/tests/test_moti.py
+++ b/tests/test_moti.py
@@ -22,34 +22,25 @@ def test_data(test_session, moti_sawr7110_xml):
     assert len(obs) - c1 == 2
 
 def test_catch_duplicates(test_session, moti_sawr7110_xml):
-    print('test_catch_duplicates')
+    """Original test of duplicates that didn't check what actually got
+    committed to the database."""
     rv = process(test_session, moti_sawr7110_xml)
     assert rv == {'failures': 0, 'successes': 2, 'skips': 2}
-    test_session.commit()
     rv = process(test_session, moti_sawr7110_xml)
     assert rv == {'failures': 0, 'successes': 0, 'skips': 4}
 
 
 def test_duplicates(test_session, moti_sawr7110_xml_2a, moti_sawr7110_xml_2b):
-    """Test that duplicates are skipped, but that skipping them does not
+    """Test that duplicates are skipped, and that skipping them does not
     cause preceding or following non-duplicate inserts to be abandoned."""
 
     # Insert initial observation set, no duplicates
-    try:
-        rv = process(test_session, moti_sawr7110_xml_2a)
-    except Exception as e:
-        assert False, \
-            "Didn't expect to get an exception on first insert\n{}".format(e)
-    # Check the claim about what happened
+    rv = process(test_session, moti_sawr7110_xml_2a)
     assert rv == {'failures': 0, 'successes': 1, 'skips': 0}
 
-    # Insert next observation set, with duplicate of first
-    try:
-        rv = process(test_session, moti_sawr7110_xml_2b)
-    except Exception as e:
-        assert False, \
-            "Didn't expect to get an exception on second insert\n{}".format(e)
-    # Check the claim about what happened
+    # Insert next observation set, with duplicate of first bracketed by
+    # non-duplicates before and after
+    rv = process(test_session, moti_sawr7110_xml_2b)
     assert rv == {'failures': 0, 'successes': 2, 'skips': 1}
 
     # Check that all non-duplicates actually got commited

--- a/tests/test_moti.py
+++ b/tests/test_moti.py
@@ -61,6 +61,9 @@ def test_duplicates(test_session, moti_sawr7110_xml_2a, moti_sawr7110_xml_2b):
         .join(Station)
         .filter(Station.native_id == '11091')
     )
+    print('Commited observations')
+    for obs in q.all():
+        print(obs.time)
     assert q.count() == 3
 
 

